building/glass/upload.py

```diff
@@ -122,20 +121,40 @@ def gs_rsync(local_path: str, remote_path: str, boto_path: str):
 """.lstrip()
 
 
-def perform_uploads(uploads: dict, branch: str) -> None:
+def upload_gs(staging, root: str, branch_config: aptbranch.Config):
+    branch = branch_config.name
+    if not GS_BRANCH_REGEX.match(branch):
+        raise Exception("not an uploadable branch: %s" % branch)
     if not os.path.exists(BOTO_PATH):
         raise Exception("you need to put the GCP service account private key file into /homeworld/boto-key")
+    botoconfig_name = os.path.join(staging, "boto.config")
+    with open(botoconfig_name, "w") as bout:
+        with open(BOTO_PATH, "r") as f:
+            project_id = json.load(f)["project_id"]
+        bout.write(BOTO_TEMPLATE % (BOTO_PATH, project_id))
+        bout.flush()
+    gs_rsync(root, branch, botoconfig_name)
+
+
+def upload_rsync(staging, root: str, branch_config: aptbranch.Config):
+    target_dir = os.path.join(branch_config.upload_config["dir"], branch_config.name)
+    target = "%s@%s:%s" % (branch_config.upload_config["username"], branch_config.upload_config["host"], target_dir)
+    subprocess.check_call(["rsync", "-av", "--progress", "--delete-delay", root + "/", target])
+
+
+def perform_uploads(uploads: dict, branch_config: aptbranch.Config) -> None:
     with tempfile.TemporaryDirectory() as staging:
         root = os.path.join(staging, "root")
         for remote_path, local_path in uploads.items():
             target = os.path.join(root, remote_path.lstrip("/"))
             if not os.path.isdir(os.path.dirname(target)):
                 os.makedirs(os.path.dirname(target))
             shutil.copy2(local_path, target)
-        botoconfig_name = os.path.join(staging, "boto.config")
-        with open(botoconfig_name, "w") as bout:
-            with open(BOTO_PATH, "r") as f:
-                project_id = json.load(f)["project_id"]
-            bout.write(BOTO_TEMPLATE % (BOTO_PATH, project_id))
-            bout.flush()
-        gs_rsync(root, branch, botoconfig_name)
+
+        upload_method = branch_config.upload_config["method"]
+        if upload_method == "google-cloud-storage":
+            upload_gs(staging, root, branch_config)
+        elif upload_method == "rsync":
+            upload_rsync(staging, root, branch_config)
```

Comment:
Maybe use a dictionary of upload methods, which maps "possible upload method" to "uploading function for that upload method"?